### PR TITLE
JSON filter 2

### PR DESCRIPTION
**Pull request by eschultink**
### Features
 - omit `type` field from schema filters
 - omit `type: array`/`type: object` where implicit
 - omit `additionalProperties`; ignored anyways

(general goal here is to improve human readability while also shrinking size to keep RULES under SSM size limits)

### Change implications

 - dependencies added/changed? **no**

### CODE REVIEW

File: java/gateway-core/src/test/java/com/avaulta/gateway/rules/JsonSchemaFilterUtilsTest.java [hunk 1]
```diff
@@ -193,21 +189,41 @@ void filterBySchema_refs() {
...
+            objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(compactCopy));
+
+        // 30% reduction in yaml size
+        assertEquals(404, yamlMapper.writeValueAsString(schemaWithRefs).length());
+        assertEquals(286, yamlMapper.writeValueAsString(compactCopy).length());
+
+        // non-pretty JSON is actually more compact than the yaml (pretty JSON is a lot less compact)
+        // (so if going to give up on human readability by base64gzip'ing the schema, might as well
+        //  do that with the JSON)
+        // eg, https://developer.hashicorp.com/terraform/language/functions/base64gzip
```
Comment (jlorper) [inline java/gateway-core/src/test/java/com/avaulta/gateway/rules/JsonSchemaFilterUtilsTest.java:216]: well, if we go this way, probably there's no point in doing all this compacting code. Just upload the full schema compressed, which make sense.
We might have special lambda calls to output rules, env variables back so can easy check config despite being compressed

**Review by jlorper — APPROVED** [review]
If we do the base64gzip then I feel all of this is irrelevant

File: java/core/src/main/java/co/worklytics/psoxy/impl/RESTApiSanitizerImpl.java [hunk 2]
```diff
@@ -92,10 +92,10 @@ Map<Endpoint, Pattern> getCompiledAllowedEndpoints() {
         return compiledAllowedEndpoints;
     }
 
-    SchemaRuleUtils.JsonSchemaFilter getRootDefinitions() {
+    JsonSchemaFilterUtils.JsonSchemaFilter getRootDefinitions() {
         if (rootDefinitions == null) {
             synchronized ($writeLock) {
-                rootDefinitions = SchemaRuleUtils.JsonSchemaFilter.builder().definitions(rules.getDefinitions()).build();
+                rootDefinitions = JsonSchemaFilterUtils.JsonSchemaFilter.builder().definitions(rules.getDefinitions()).build();
```
Comment (jlorper) [inline java/core/src/main/java/co/worklytics/psoxy/impl/RESTApiSanitizerImpl.java:98]: for doing the double locking singleton properly you need to check again in the lock

```suggestion
  if (rootDefinitions == null) {
          rootDefinitions = JsonSchemaFilterUtils.JsonSchemaFilter.builder().definitions(rules.getDefinitions()).build();
  }
```

File: java/gateway-core/src/test/java/com/avaulta/gateway/rules/JsonSchemaFilterUtilsTest.java (hunk 1 above)
Comment (eschultink) [reply java/gateway-core/src/test/java/com/avaulta/gateway/rules/JsonSchemaFilterUtilsTest.java:220]: yeah, i prefer to keep human readable if we can, so people can just see the value in their SSM parameters to have confidence in it.  It's possible base64gzip would be conditional just for the extreme cases (eg, if > 8KB compacted it then gzip)

**Pull request merged**